sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp

```diff
@@ -0,0 +1,679 @@
+
+  if (!acro_form_dictionary->KeyExist("XFA")) {
+    return false;
+  }
+
+  CPDF_Object* pdf_xfa_object = acro_form_dictionary->GetElement("XFA");
+  CPDF_Array* pdf_xfa_array = pdf_xfa_object->GetArray();
+  if (!pdf_xfa_array) return false;
+
+  FX_DWORD count = pdf_xfa_array->GetCount();
+  for (int i = 0; i < static_cast<int>(count); i++) {
+    CPDF_Object* pdf_object = pdf_xfa_array->GetElement(i);
+    if (NULL != pdf_object && pdf_object->GetType() == PDFOBJ_REFERENCE) {
+      CPDF_Object* pdf_reference_object = (CPDF_Object*)pdf_object->GetDict();
+      if (NULL != pdf_reference_object) {
+        CPDF_Dictionary* pdf_dictionary = pdf_reference_object->GetDict();
+        if (NULL != pdf_dictionary && pdf_dictionary->GetInteger("Length") != 0) {
+          if (root_dictionary->KeyExist("NeedsRendering")) {
+            bool dynamic_xfa = (root_dictionary->GetBoolean("NeedsRendering") != 0);
+            if (dynamic_xfa) {
+              return true;
+            } else {
+              //is static XFA
+              return false;
+            }
+          } else {
+            //is static XFA
+            return false;
+          }
+        }
+      }
+    }
+  }
+  return false;
+}
+
+uint32_t PDFCreatorImpl::ParsePDFFile(CPDF_Parser *pdf_parser) {
+  uint32_t result = PDFCreatorErr::SUCCESS;
+  CFX_ByteString file_path_bytestring = file_path_.c_str();
+
+  FX_DWORD parse_result = pdf_parser->StartParse(file_path_bytestring);
+  result = ConvertParsingErrCode(parse_result);
+
+  if(result == PDFCreatorErr::SUCCESS) {
+    //has been protected by password
+    if (IsProtectedByPassword(pdf_parser)) {
+      return PDFCreatorErr::SECURITY;
+    }
+
+    //has been signed?
+    if (IsSigned(pdf_parser)) {
+      return PDFCreatorErr::FORMAT;
+    }
+
+    //is dynamic XFA PDF
+    if (IsDynamicXFA(pdf_parser)) {
+      return PDFCreatorErr::FORMAT;
+    }
+  }
+
+  return result;
+}
+
+uint32_t PDFCreatorImpl::ConvertParsingErrCode(FX_DWORD parse_result) {
+  uint32_t result = PDFCreatorErr::SUCCESS;
+  switch (parse_result) {
+    case PDFPARSE_ERROR_SUCCESS: {
+      result = PDFCreatorErr::SUCCESS;
+      break;
+    }
+    case PDFPARSE_ERROR_FILE: {
+      result = PDFCreatorErr::FILE;
+      break;
+    }
+    case PDFPARSE_ERROR_FORMAT: {
+      result = PDFCreatorErr::FORMAT;
+      break;
+    }
+
+    case PDFPARSE_ERROR_PASSWORD:
+    case PDFPARSE_ERROR_HANDLER:
+    case PDFPARSE_ERROR_CERT: {
+      result = PDFCreatorErr::SECURITY;
+      break;
+    }
+
+    default: {
+      result = PDFCreatorErr::UNKNOWN_ERR;
+      break;
+    }
+
+  }
+
+  return result;
+}
+
+CPDF_Dictionary* PDFCreatorImpl::CreateEncryptionDict(
+    const std::string& filter_name,
+    const std::vector<unsigned char> &publishing_license) {
+  CPDF_Dictionary* encryption_dictionary = new CPDF_Dictionary;
+
+  encryption_dictionary->SetAtName("Filter", filter_name.c_str());
+
+  CPDF_Dictionary* pCFDic = new CPDF_Dictionary;
```

Comment:
Where is this memory deallocated?

Reply:
pCFDic is added to encryption_dictionary. pCFDic will be released in encryption_dictionary's destructor function CPDF_Dictionary::~CPDF_Dictionary(). encryption_dictionary will be released in CPDF_Parser::~CPDF_Parser().